crates/holochain_p2p/src/spawn/actor.rs

```diff
@@ -638,14 +638,29 @@ impl kitsune_p2p::event::KitsuneP2pEventHandler for HolochainP2pActor {
         &mut self,
         input: kitsune_p2p::event::PutAgentInfoSignedEvt,
     ) -> kitsune_p2p::event::KitsuneP2pEventHandlerResult<()> {
-        let kitsune_p2p::event::PutAgentInfoSignedEvt { space, peer_data } = input;
-        let space = DnaHash::from_kitsune(&space);
+        let kitsune_p2p::event::PutAgentInfoSignedEvt { peer_data } = input;
+
         let evt_sender = self.evt_sender.clone();
-        Ok(
-            async move { Ok(evt_sender.put_agent_info_signed(space, peer_data).await?) }
-                .boxed()
-                .into(),
-        )
+        Ok(async move {
+            let put_requests = peer_data
+                .into_iter()
+                .map(|agent| (DnaHash::from_kitsune(&agent.space), agent))
+                .fold(
+                    HashMap::<DnaHash, Vec<AgentInfoSigned>>::new(),
+                    |mut acc, (dna, agent)| {
+                        acc.entry(dna).or_default().push(agent);
+                        acc
+                    },
+                );
+
+            for (dna, agents) in put_requests {
+                evt_sender.put_agent_info_signed(dna, agents).await?;
```

Comment:
Can putting be done in parallel if that makes sense?

Reply:
It can, just requires a rather ugly block of code - done :)

Reply:
Amazingly ugly indeed!